ToolBox/TradingEconomicsDataDownloader/TradingEconomicsIndicatorDownloader.cs

```diff
@@ -67,44 +98,94 @@ public override bool Run()
+                            startUtc = startUtc.AddMonths(1);
+                            continue;
+                        }
+
+                        Log.Trace($"TradingEconomicsIndicatorDownloader.Run(): Collecting data for indicator: {indicator} - from {startUtc:yyyy-MM-dd} to {endUtc:yyyy-MM-dd}");
+
+                        _requestGate.WaitToProceed(TimeSpan.FromSeconds(1));
+
                         var content = Get(startUtc, endUtc).Result;
                         var collection = JsonConvert.DeserializeObject<List<TradingEconomicsIndicator>>(content);
 
                         data.AddRange(collection);
-
                         startUtc = startUtc.AddMonths(1);
                     }
                     catch (Exception e)
                     {
-                        Log.Error(e, $"TradingEconomicsIndicatorsDownloader(): Error parsing data for date {startUtc:yyyyMMdd}");
+                        Log.Error(e, $"TradingEconomicsIndicatorDownloader.Run(): Error parsing data for date {startUtc:yyyyMMdd}");
                         return false;
                     }
                 }
 
-                Log.Trace($"TradingEconomicsIndicatorsDownloader(): {data.Count} {indicator} indicator entries read in {stopwatch.Elapsed}");
+                Log.Trace($"TradingEconomicsIndicatorDownloader.Run(): {data.Count} {indicator} indicator entries read in {stopwatch.Elapsed}");
 
-                foreach (var kvp in data.GroupBy(GetFileName))
-                {
-                    var path = Path.Combine(_destinationFolder, kvp.Key);
-                    var zipPath = path.Replace(".json", ".zip");
+                // Return status code. We default to `true` so that we can identify if an error occured during the loop
+                var status = true;
 
-                    try
-                    {
-                        var contents = JsonConvert.SerializeObject(kvp.ToList());
-                        File.WriteAllText(path, contents);
-                        // Write out this data string to a zip file
-                        Compression.Zip(path, zipPath, kvp.Key, true);
-                    }
-                    catch (Exception e)
+                Parallel.ForEach(data.GroupBy(x => GetTicker(x.HistoricalDataSymbol, x.Category, x.Country)),
+                    (kvp, state) =>
                     {
-                        Log.Error(e, $"TradingEconomicsIndicatorsDownloader(): Error creating {path}");
-                        return false;
+                        // Create the destination directory, otherwise we risk having it fail when we move
+                        // the temp file to its final destination
+                        Directory.CreateDirectory(Path.Combine(_destinationFolder, kvp.Key));
+
+                        foreach (var indicatorByDate in kvp.GroupBy(x => x.LastUpdate))
+                        {
+                            var date = indicatorByDate.Key.ToString("yyyyMMdd");
```

Comment:
In cases like this is preferable to use `indicatorByDate.Key.ToString(DateFormat.EightCharacter)`.